File: 328. Odd Even Linked List.cpp
// !!!!!OMG I DID IT!!!!! I CANNOT BELIEVE I DID IT!!!
// I ACed at the very first try!!!!! UNBELIEVABLE!!!
// It's a linked list involving so many operations!!!
// I'm on a roll right now...
// Time: O(n), space: O(1)
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        if (!head || !(head->next)) return head;
        ListNode *odd = head, *even = head -> next;
        ListNode *cur = even -> next;
        while (cur) {
            ListNode *oddnext = odd -> next;
            odd -> next = cur;
            even -> next = cur -> next;
            even = even -> next;
            cur -> next = oddnext;
            odd = cur;
            if (even) cur = even -> next;
            else break;
        }
        
        return head;
    }
};

File: 328. Odd Even Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def oddEvenList(self, head: ListNode) -> ListNode:
        if not head or not head.next or not head.next.next: return head
        odd, even, cur = head, head.next, head.next.next
        while even and cur:
            even.next = cur.next
            cur.next = odd.next
            odd.next = cur
            odd = cur
            even = even.next
            if even: cur = even.next
        return head
